pkg/app/api/grpcapi/piped_api.go

```diff
@@ -679,6 +681,94 @@ func (a *PipedAPI) ReportApplicationLiveStateEvents(ctx context.Context, req *pi
 	return &pipedservice.ReportApplicationLiveStateEventsResponse{}, nil
 }
 
+// GetLatestEvent gives back the latest event that meets the given conditions.
+func (a *PipedAPI) GetLatestEvent(ctx context.Context, req *pipedservice.GetLatestEventRequest) (*pipedservice.GetLatestEventResponse, error) {
+	projectID, _, _, err := rpcauth.ExtractPipedToken(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Try to fetch the most recently registered event.
+	opts := datastore.ListOptions{
+		PageSize: 1,
+		Filters: []datastore.ListFilter{
+			{
+				Field:    "ProjectId",
+				Operator: "==",
+				Value:    projectID,
+			},
+			{
+				Field:    "Name",
+				Operator: "==",
+				Value:    req.Name,
+			},
+			// TODO: Enable to use labels to filter Events
+		},
+		Orders: []datastore.Order{
+			{
+				Field:     "UpdatedAt",
+				Direction: datastore.Desc,
+			},
+		},
+	}
+	events, err := a.eventStore.ListEvents(ctx, opts)
+	if err != nil {
+		a.logger.Error("failed to list events", zap.Error(err))
+		return nil, status.Error(codes.Internal, "failed to list event")
+	}
+	if len(events) == 0 {
+		a.logger.Error("no events found",
+			zap.String("project-id", projectID),
+			zap.String("name", req.Name),
+			zap.Any("labels", req.Labels),
+		)
+		return nil, status.Error(codes.NotFound, "no events found")
+	}
+
+	return &pipedservice.GetLatestEventResponse{
+		Event: events[0],
+	}, nil
+}
+
+// ListEvents returns a list of Events newly created after the given milestone.
+func (a *PipedAPI) ListEvents(ctx context.Context, req *pipedservice.ListEventsRequest) (*pipedservice.ListEventsResponse, error) {
+	projectID, _, _, err := rpcauth.ExtractPipedToken(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	opts := datastore.ListOptions{
+		Filters: []datastore.ListFilter{
+			{
+				Field:    "ProjectId",
+				Operator: "==",
+				Value:    projectID,
+			},
+			{
+				Field:    "CreatedAt",
+				Operator: ">",
+				Value:    req.Milestone,
+			},
+		},
+	}
+
+	events, err := a.eventStore.ListEvents(ctx, opts)
+	if err != nil {
+		a.logger.Error("failed to list events", zap.Error(err))
+		return nil, status.Error(codes.Internal, "failed to list events")
+	}
+	if len(events) == 0 {
+		a.logger.Error("no events found",
+			zap.String("project-id", projectID),
+			zap.Int64("milestone", req.Milestone),
+		)
+		return nil, status.Error(codes.NotFound, "no events found")
+	}
```

Comment:
Other List~ API doesn't treat the not found case as an error, but I'd say it should be handled.

Reply:
I am not sure why we need to add this NotFound error while the empty list already means nothing was found.

Reply:
Come to think of it, right, it represents Not Found well. I'll fix it.